modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -98,265 +101,291 @@
+
+    /** Constructor. */
+    ReliableChannel(ClientConfiguration clientCfg,
+                    Function<ClientChannelConfiguration, ClientChannel> chFactory,
+                    boolean initAllChannels) {
         if (chFactory == null)
             throw new NullPointerException("chFactory");
 
         if (clientCfg == null)
             throw new NullPointerException("clientCfg");
 
+        this.clientCfg = clientCfg;
         this.chFactory = chFactory;
+        this.initAllChannels = initAllChannels;
+    }
+
+    /** Should the channel initialization be stopped. */
+    private boolean stopInitCondition() {
+        return scheduledChannelsReinit.get() || closed;
+    }
 
-        List<InetSocketAddress> addrs = parseAddresses(clientCfg.getAddresses());
+    /** Callback is invoked after new ClientChannel has created. */
+    private final BiConsumer<ClientChannelHolder, ClientChannel> onChannelCreate = (holder, ch) -> {
+        ch.addTopologyChangeListener(channel -> {
+            if (topChangeFilters.stream().allMatch(s -> s.test(channel)))
+                channelsInit(true);
+        });
 
-        channels = new ClientChannelHolder[addrs.size()];
+        ch.addNotificationListener(this);
 
-        for (int i = 0; i < channels.length; i++)
-            channels[i] = new ClientChannelHolder(new ClientChannelConfiguration(clientCfg, addrs.get(i)));
+        nodeChannels.values().remove(holder);
+        nodeChannels.put(ch.serverNodeId(), holder);
+    };
 
-        curChIdx = new Random().nextInt(channels.length); // We already verified there is at least one address.
+    /** Callback is invoked after a ClientChannel has closed. */
+    private final Consumer<ClientChannel> onChannelClose = ch -> {
+        for (Consumer<ClientChannel> lsnr : channelCloseLsnrs)
+            lsnr.accept(ch);
+    };
 
-        partitionAwarenessEnabled = clientCfg.isPartitionAwarenessEnabled() && channels.length > 1;
+    /**
+     * Init channel holders to all nodes.
+     * @param force enable to replace existing channels with new holders.
+     */
+    private synchronized void initChannelHolders(boolean force) {
+        // enable parallel threads to schedule new init of channel holders
+        scheduledChannelsReinit.set(false);
+
+        if (!force && channels.get() != null)
+            return;
+
+        List<InetSocketAddress> resolvedAddrs = parseAddresses(clientCfg.getAddresses());
+
+        List<ClientChannelHolder> holders = Optional.ofNullable(channels.get()).orElse(new ArrayList<>());
+
+        // addr -> (holder, delete)
+        Map<InetSocketAddress, T2<ClientChannelHolder, Boolean>> addrs = holders.stream()
+            .collect(Collectors.toMap(
+                c -> c.chCfg.getAddress(),
+                c -> new T2<>(c, null)
+        ));
+
+        // mark for delete addrs that aren't provided by clientConfig now
+        addrs.keySet()
+            .stream()
+            .filter(addr -> !resolvedAddrs.contains(addr))
```

Comment:
`resolvedAddrs` is a List, this can be inefficient with large number of channels and addresses

Reply:
replaced  with Set

Reply:
I'm not sure that replacing `List` to `Set` it's a good idea here (perhaps you should just copy list to set here locally). Using `List` you can solve some user problems which can't be solved with `Set`. For example, using `list` you can make some server nodes to be used more often than another. Also, by having a single duplicated address you can workaround some known issues.

Reply:
Fixed, add tests for that.